### PR TITLE
[feature] Crowding widget backend

### PR DESCRIPTION
**Pull request by hannahpurcell**
**Asana task**: [OL Crowding: start backend logic](https://app.asana.com/0/1185117109217413/1205185788948508)

And also [this task](https://app.asana.com/0/1185117109217422/1205279706586348)

Tests are in-progress.  This covers the backend logic that determines the conditions for showing the crowding widget, and serializes it.  

Some of the changes are just renaming, or adding "Triptych" to an array of app names

- [x] Tests added?

### CODE REVIEW

File: lib/screens/vehicles/parser.ex [hunk 1]
```diff
@@ -19,13 +20,28 @@ defmodule Screens.Vehicles.Parser do
...
 
+  defp parse_carriages(data), do: Enum.map(data, &parse_car_crowding/1)
+
+  defp parse_car_crowding(%{
+         "occupancy_status" => occupancy_status,
+         "occupancy_percentage" => occupancy_percentage,
+         "carriage_sequence" => carriage_sequence
+       }) do
+    %{
+      occupancy_status: parse_occupancy_status(occupancy_status),
```
Comment (cmaddox5) [inline lib/screens/vehicles/parser.ex:39]: I believe OL crowding uses a different occupancy level system than bus, so we may need to add a `parse_occupancy_status` for just OL. If you look [here](https://www.notion.so/mbta-downtown-crossing/Orange-Line-Car-Crowding-UX-Design-7e31f94cb0ab47eeacd3489877a040d4?pvs=4#abb0744aafe1471e960d34f3d7049ea2), the API uses 5 levels and Betsy created this table for converting the 5 down to 3.

File: lib/screens/v2/widget_instance/train_crowding.ex [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+          destination: String.t(),
+          arrival_time: String.t(),
+          crowding: list(crowding_data),
+          platform_position: number,
+          front_car_direction: :left | :right,
+          now: String.t()
+        }
+
+  @type crowding_data :: %{
+          crowding_level: 1 | 2 | 3,
```
Comment (cmaddox5) [inline lib/screens/v2/widget_instance/train_crowding.ex:30]: Should these values be the level atoms from the vehicle parser? `:many_seats_available`, `:full`, etc.

File: lib/screens/v2/widget_instance/train_crowding.ex [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+          percentage: number
```
Comment (cmaddox5) [inline lib/screens/v2/widget_instance/train_crowding.ex:31]: To make things super simple, we should be able to remove this. If I'm understanding correctly, we will always pick the crowding color on the frontend based on `crowding_level`. Should need `percentage` at all.

File: lib/screens/v2/widget_instance/train_crowding.ex (hunk 2 above)
Comment (cmaddox5) [reply lib/screens/v2/widget_instance/train_crowding.ex:30]: Also, we will need to add some values for closed cars and no data. I'm not 100% what that looks like in the data though. I asked [here](https://mbta.slack.com/archives/C039ARUM48H/p1692880405163159?thread_ts=1692710156.845019&cid=C039ARUM48H).

File: lib/screens/v2/widget_instance/train_crowding.ex [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+  @spec serialize(t()) :: widget_data()
+  def serialize(%__MODULE__{
+        screen: %Screen{app_params: %Triptych{train_crowding: train_crowding}},
+        prediction: prediction,
+        now: now
+      }) do
+    %{
+      destination: prediction.trip.headsign,
+      arrival_time: serialize_time(prediction.arrival_time),
```
Comment (cmaddox5) [inline lib/screens/v2/widget_instance/train_crowding.ex:42]: Can this be omitted? I don't see us needing this in the client.

**Review by cmaddox5 — COMMENTED** [review]
So sorry for the delay! Have a few questions, suggestions, and comments.

File: lib/screens/v2/candidate_generator/widgets/train_crowding.ex [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    with {:ok, predictions} <- fetch_predictions_fn.(params),
+         {:ok, location_context} <-
+           fetch_location_context_fn.(Triptych, train_crowding.station_id, now),
+         {:ok, alerts} <-
+           params |> Map.to_list() |> fetch_alerts_fn.() do
+      next_train_prediction = List.first(predictions)
+
+      # If there is an upcoming train, it's headed to this station, and we're not at a temporary terminal,
+      # show the widget
+      if !is_nil(next_train_prediction) and
```
[the remark this answers is not in the document]
Comment (hannahpurcell) [reply lib/screens/v2/candidate_generator/widgets/train_crowding.ex:56]: There are 4 conditions here that need to be met for the widget instance stuff, and if any one of them is false, we do the empty array. So I could switch to this, but it seems less clear:

```
      cond do
        is_nil(next_train_prediction) -> []
        Prediction.vehicle_status(next_train_prediction) != :incoming_at -> []
        next_train_prediction |> Prediction.stop_for_vehicle() |> fetch_parent_stop_id_fn.() !=
            train_crowding.station_id -> []
        any_alert_makes_this_a_terminal?(alerts, location_context) -> []
        true -> [
              %CrowdingWidget{
                screen: config,
                prediction: next_train_prediction,
                now: now
              }
            ]
      end
```

File: lib/screens/v2/widget_instance/train_crowding.ex (hunk 4 above)
Comment (hannahpurcell) [reply lib/screens/v2/widget_instance/train_crowding.ex:42]: Oh yes! ++ We did need it at one point when the logic was "if the train is 2 minutes away", but that's old now.

File: lib/screens/v2/widget_instance/train_crowding.ex (hunk 3 above)
Comment (hannahpurcell) [reply lib/screens/v2/widget_instance/train_crowding.ex:31]: Oh yeah! Seems so!

File: lib/screens/v2/widget_instance/train_crowding.ex (hunk 2 above)
Comment (hannahpurcell) [reply lib/screens/v2/widget_instance/train_crowding.ex]: Ah yep, this whole type was old.

File: lib/screens/v2/candidate_generator/widgets/train_crowding.ex (hunk 5 above)
[the remark this answers is not in the document]
Comment (cmaddox5) [reply lib/screens/v2/candidate_generator/widgets/train_crowding.ex]: Ehh, I agree. That's hard to follow. You can just ignore my initial comment. I misunderstood the conditional when I wrote that.

File: lib/screens/v2/widget_instance/train_crowding.ex [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+      destination: prediction.trip.headsign,
+      crowding: serialize_carriages(prediction.vehicle.carriages),
+      platform_position: train_crowding.platform_position,
+      front_car_direction: train_crowding.front_car_direction,
+      now: serialize_time(now)
+    }
+  end
+
+  defp serialize_time(%DateTime{} = time) do
+    DateTime.to_iso8601(time)
```
Comment (cmaddox5) [inline lib/screens/v2/widget_instance/train_crowding.ex:46]: Does this need to be timezone shifted? Or were you thinking that can be done on the client?

File: lib/screens/v2/widget_instance/train_crowding.ex [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+  defp serialize_carriages(carriages),
+    do: Enum.map(carriages, fn car -> serialize_occupancy_status(car) end)
+
+  defp serialize_occupancy_status(:no_data_available), do: :no_data
+  defp serialize_occupancy_status(:many_seats_available), do: :not_crowded
+  defp serialize_occupancy_status(:few_seats_available), do: :not_crowded
+  defp serialize_occupancy_status(:standing_room_only), do: :some_crowding
+  defp serialize_occupancy_status(:crushed_standing_room_only), do: :crowded
+  defp serialize_occupancy_status(:full), do: :crowded
+  defp serialize_occupancy_status(:not_accepting_passengers), do: :disabled
```
Comment (cmaddox5) [inline lib/screens/v2/widget_instance/train_crowding.ex:60]: Nitpicky, but can this conversion live in the vehicle parser? I would imagine if we were to show crowding info for other lines, it would use the same conversion system. Total guess though so up to you.

**Review by cmaddox5 — APPROVED** [review]
A couple of questions, but this looks great!

File: lib/screens/v2/widget_instance/train_crowding.ex (hunk 7 above)
Comment (hannahpurcell) [reply lib/screens/v2/widget_instance/train_crowding.ex:60]: Hmm. In my comments to your previous comments, I mentioned my rationale here is that parsing should be a direct translation, while serialize handles "present it this way in this particular context". I can see us using this conversion in other situations, but I can also see us not doing that, and choosing to use all 5 crowding types. Maybe I'll ask design

File: lib/screens/v2/widget_instance/train_crowding.ex (hunk 6 above)
Comment (hannahpurcell) [reply lib/screens/v2/widget_instance/train_crowding.ex:46]: I mimicked the serializer for AlertHeader. And it seems to be translating correctly with the <Time> component on the frontend, so I think we're good

File: lib/screens/v2/candidate_generator/widgets/train_crowding.ex [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+           params |> Map.to_list() |> fetch_alerts_fn.() do
+      next_train_prediction = List.first(predictions)
+
+      # If there is an upcoming train, it's headed to this station, and we're not at a temporary terminal,
+      # show the widget
+      if not is_nil(next_train_prediction) and
+           Prediction.vehicle_status(next_train_prediction) == :incoming_at and
+           next_train_prediction |> Prediction.stop_for_vehicle() |> fetch_parent_stop_id_fn.() ==
+             train_crowding.station_id and
+           not any_alert_makes_this_a_terminal?(alerts, location_context) do
```
Comment (hannahpurcell) [inline lib/screens/v2/candidate_generator/widgets/train_crowding.ex:60]:
```suggestion
           next_train_prediction.vehicle.carriages != [] and
           not any_alert_makes_this_a_terminal?(alerts, location_context) do
```

**Review by cmaddox5 — APPROVED** [review]
🚀
